### DISCUSSION TITLE
Too many requests, response 429, Retry-After header access

### DISCUSSION THREAD

**Discussion by AurJur** [spotify-web-api-java/spotify-web-api-java#362] [General]
Recently (about a month ago) I started getting "Too Many Requests" exception in places where it had not occurred earlier. Has someone noticed it? Did Spotify change something (made limits lower) recently?
I noticed one of the recent commits in this project is related to 429. @SergioBarbero How does it affect the issue? I am still on 8.0.0 version. [Releases page](https://github.com/spotify-web-api-java/spotify-web-api-java/releases) is only up to v8.0.0, no notes on later versions.
Spotify docs mention a header Retry-After which should give value in seconds for a pause before a next request. Is this implemented here? In general, how to access headers using this API?
Thanks.

**Comment by dargmuesli** [comment]
@hwup might be quicker to answer questions related to the 429 change, I'm currently only maintaining this wrapper, not using it much myself. Reason for the mentioned change was issue https://github.com/spotify-web-api-java/spotify-web-api-java/issues/344. I think that instead of having a freeze, a request returning 429 is now throwing an exception. @hwup could you confirm or give insight into the other questions if you've been using the code recently?

I've added the missing release notes now.

**Comment by thisisjonchen** [comment]
Just happened to get "Too Many Requests" exception only on Get Playlists starting yesterday; fine on all others. Worked perfectly fine before. Also had the freeze previously, but updated to v8.3.0, and that seemed to fix it, but it was replaced with 429.

**Comment by hwup** [comment]
Previously when it encountered 429, it would sleep for amount of `Retry-After` header, so you would not notice when you would get 429 and it was a short wait, in case it's a long wait, you would notice because it would "hang". After the PR https://github.com/spotify-web-api-java/spotify-web-api-java/pull/345, now it throws exception and you can choose what to do. You can use `getRetryAfter` in `ToomanyRequestsException` and sleep, etc.

> Spotify docs mention a header Retry-After which should give value in seconds for a pause before a next request. Is this implemented here? In general, how to access headers using this API? Thanks.
https://github.com/spotify-web-api-java/spotify-web-api-java/blob/0799b425904494c50cf874c66d7f92a48099403a/src/main/java/se/michaelthelin/spotify/exceptions/detailed/TooManyRequestsException.java#L8-L11
